river/anomaly/xstream.py

```diff
@@ -0,0 +1,251 @@
+import math
+import random
+from collections import Counter
+
+from river import anomaly
+from river.transform.projection.streamhash_projector import StreamhashProjector
+from river.utils.math import dict_zeros, get_minmax, merge
+
+random.seed(14)
+
+
+class xStream(anomaly.base.AnomalyDetector):
+    """The xStream model for row-streaming data :cite:`xstream`. It first projects the data via streamhash projection. It then fits half space chains by reference windowing. It scores the instances using the window fitted to the reference window.
+
+    Parameters
+    ----------
+    num_components
+        Number of components for streamhash projection (Default=100).
+    n_chains
+        Number of half-space chains (Default=100).
+    depth
+        Maximum depth for the chains (Default=25).
+    window_size
+        Size (and the sliding length) of the reference window (Default=25).
+
+    """
+
+    def __init__(self, num_components=100, n_chains=100, depth=25, window_size=25):
+        self.streamhash = StreamhashProjector(num_components=num_components)
+        deltamax = {}
+        for i in range(num_components):
+            deltamax[i] = 0.5
+        deltamax = {
+            feature: 1.0 if (value <= 0.0001) else value for feature, value in deltamax.items()
+        }
+        self.window_size = window_size
+        self.hs_chains = _HSChains(deltamax=deltamax, n_chains=n_chains, depth=depth)
+
+        self.step = 0
+        self.cur_window = []
+        self.ref_window = None
+
+    def learn_one(self, x, y=None):
+        """Fits the model to next instance.
+
+        Parameters
+        ----------
+        X
+            Instance to learn.
+        y
+            Ignored since the model is unsupervised (Default=None).
+
+        """
+        self.step += 1
+
+        X = self.streamhash.learn_one(x)
+        X = self.streamhash.transform_one(X)
+        self.cur_window.append(X)
+        self.hs_chains.fit(X)
```

Comment:
Is there no way to train `hs_chains` incrementally? Does it have `fit` on all of `X` each time a new sample is seen?